docs/specs/priority-queue.md

```diff
@@ -1,70 +1,99 @@
-queues:
+### Design decision: smallest priority first
 
-* `PriorityQueue<T>`,
-* `PriorityQueue<TElement, TPriority>`.
+This priority queue puts the item with minimum priority first, i.e. 'smallest' priority value is 'highest' priority.
 
-## `PriorityQueue<T>`
+### Design decision: mutability of priorities, and notification of priority changes 
+
+Priorities are *mutable* so that the data structure is easy to use for popular scenarios such as shortest-path algorithms. By possible scenarios:
+
+1. Extrinsic priorities can be updated by assigning the new priority value, like a dictionary. For consistency and correctness it doesn't matter whether updates are done one at a time, or in a batch.
+2. Intrinsic priorities sometimes change over time. It is the responsibility of whoever causes the change to *notify* the priority queue about *all* changes which happen so that the data structure can be kept consistent and up to date.
+
+Priorities are updatable (like dictionary values are updatable). 
+
+That priority value may change over time.
+
+#### Workarounds for the uniqueness design decision
+
+
+
+### Modelling priority changes (single and multiple)
+
+For items with *extrinsic* priority, the priority is updated by explicitly updating the priority of the item with a new value, using the dictionary setter. In this case the data structure can be maintained in correct order whether item priorities change simultaneously, or in sequence.
+For items with *intrinsic* priority, the priority is updated due to some kind of change in the item itself, which will cause the results of its IComparable<> comparison to change relative to other items.
+Whenever such a change to an item happens, the priority queue just be informed in a single call of the list of ALL items whose priority has changed, to ensure that no data-structure invariants are violated when the update-priority algorithmm is executed.
+
+## The proposed classes
+
+There are actually two PriorityQueue classes, one to support each scenario. PriorityQueue<TItem, TPriority> supports the *extrinsic priority* scenario, and PrioritySet<T> supports the *intrinsic priority* scenario. 
+
+## `PrioritySet<T>` (for intrinsic priority)
+
+`PrioritySet<T>` is a set of items with intrinsic priorities. It must be initialized with a 'priority comparer' which understands how to compare the intrinsic priorities of each item.
+
+It differs from `SortedSet<T>` because it allows separate items to have equal priorities, and because it does not maintain strict ordering at all times, and because it allows that priorities are mutable.
+
+If you provide `DefaultComparer<T>` as your priority comparer, and never mutate priorities, you will get similar behavior to `SortedSet<T>`, where priorities are the items and priorities are all distinct.
 
 ```csharp
-public class PriorityQueue<T> : IQueue<T>
+public class PrioritySet<T> : ISet<T>, ICollection<T>, IEnumerable<T>, IEnumerable, 
```

Comment:
nit: class name is now `PrioritySet` but the ctor names are not. I would prefer queue and follow c++ `priority_queue` inspired design http://en.cppreference.com/w/cpp/container/priority_queue

Reply:
And I don't think `ISet<T>` is appropriate, what wil `IsSubsetOf` mean? etc.